numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -679,15 +679,62 @@ NPY_NO_EXPORT npy_bool
 PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
                                                     NPY_CASTING casting)
 {
-    /* Fast path for unsafe casts or basic types */
-    if (casting == NPY_UNSAFE_CASTING ||
-            (NPY_LIKELY(from->type_num < NPY_OBJECT) &&
+    /*
+     * Fast path for basic types.
+     */
+    if ((NPY_LIKELY(from->type_num < NPY_OBJECT) &&
              NPY_LIKELY(from->type_num == to->type_num) &&
              NPY_LIKELY(from->byteorder == to->byteorder))) {
         return 1;
     }
-    /* Equivalent types can be cast with any value of 'casting'  */
-    else if (PyArray_EquivTypenums(from->type_num, to->type_num)) {
+    /*
+     * Cases with fields need special treatment.
+     */
+    if (PyDataType_HASFIELDS(from)) {
+        /*
+         * If from is a structured data type, then it can be cast to a simple
+         * non-object one only for unsafe casting *and* if it has a single field.
+         */
+        if (!PyDataType_HASFIELDS(to) && to->type_num != NPY_OBJECT) {
+            return (casting == NPY_UNSAFE_CASTING &&
+                    PyDict_Size(from->fields) == 1);
```

Comment:
I wonder if Eric won't come around a corner here and spook you with a dtype with nested fields :), it seems probable that there needs to be a recursion somewhere to get it right.

Reply:
You're right, this should be recursive.